publish-mqtt/src/main.rs

```diff
@@ -291,78 +383,87 @@ async fn disconnect_first_stale_sensor(
+    bt_session: MijiaSession,
+) -> Result<(), anyhow::Error> {
+    println!("Subscribing to events");
+    let (msg_match, mut events) = bt_session.event_stream().await?;
+    println!("Processing events");
     // Process events until there are none available for the timeout.
-    for event in bt_session
-        .incoming(INCOMING_TIMEOUT_MS)
-        .filter_map(BluetoothEvent::from)
-    {
-        handle_bluetooth_event(homie, sensors_connected, sensors_to_connect, event).await?;
+    while let Some(event) = events.next().await {
+        handle_bluetooth_event(state.clone(), event)
+            .await
+            .with_context(|| std::line!().to_string())?
     }
-    Ok(())
+    bt_session
+        .connection
+        .remove_match(msg_match.token())
+        .await?;
+    panic!("no more events");
 }
 
 async fn handle_bluetooth_event(
-    homie: &mut HomieDevice,
-    sensors_connected: &mut Vec<Sensor>,
-    sensors_to_connect: &mut VecDeque<Sensor>,
-    event: BluetoothEvent,
-) -> Result<(), Box<dyn Error>> {
+    state: Arc<Mutex<SensorState>>,
+    event: MijiaEvent,
+) -> Result<(), anyhow::Error> {
+    let state = &mut *state.lock().await;
+    let homie = &mut state.homie;
+    let sensors_connected = &mut state.sensors_connected;
+    let sensors_to_connect = &mut state.sensors_to_connect;
     match event {
-        BluetoothEvent::Value { object_path, value } => {
-            // TODO: Make this less hacky.
-            let device_path = match object_path.strip_suffix(SERVICE_CHARACTERISTIC_PATH) {
-                Some(path) => path,
-                None => return Ok(()),
-            };
+        MijiaEvent::Readings {
+            object_path,
+            readings,
+        } => {
             if let Some(sensor) = sensors_connected
                 .iter_mut()
-                .find(|s| s.device_path == device_path)
+                .find(|s| s.object_path == object_path)
             {
-                sensor.last_update_timestamp = Instant::now();
-                if let Some(readings) = decode_value(&value) {
-                    sensor.publish_readings(homie, &readings).await?;
-                } else {
-                    println!("Invalid value from {}", sensor.name);
-                }
-            } else {
-                // TODO: Still send it, in case it is useful?
-                println!("Got update from unexpected device {}", device_path);
+                sensor.publish_readings(homie, &readings).await?;
+            } else if let Some(sensor_index) = sensors_to_connect
+                .iter()
+                .position(|s| s.object_path == object_path)
+            {
+                let mut sensor = sensors_to_connect.remove(sensor_index).unwrap();
+                println!(
+                    "Got update from disconnected device {}. Connecting.",
+                    object_path
+                );
+                sensor.publish_readings(homie, &readings).await?;
```

Comment:
You first need to call `homie.add_node`. You may also want to update the  `sensor.connection_status`.